Support --cpuset-cpus flag with docker create

shohamd4 opened the pull request.

JaimeFrey committed:
(#7427) Make new calls to [fd]PrintAd() in v8.9 exclude private attributes

Sam-Tygier committed:
In condor_prio manpage make ordering more prominent

Make the ordering of priority slightly more prominent

Good to have this clear, as it is inverted compared to user priority
and UNIX nice values.

Sam-Tygier committed:
In condor_prio manpage make ordering more prominent

Make the ordering of priority slightly more prominent

Good to have this clear, as it is inverted compared to user priority
and UNIX nice values.

GregThain committed:
Explicitly ignore return value when it doesn't matter #6992

GregThain committed:
Add first cut at README.md for github

Todd-L-Miller committed:
Add a second cut to the README.md file.

GregThain committed:
Check return value from system calls #6992

johnkn committed:
fix collection iter mismatch in consumption policies #6992

zmiller committed:
Rename user-visible filenames and config names for consistency and to match upcoming future OAuth commits.  #7376

GregThain committed:
Don't refer nullptrs #6992

GregThain committed:
Check for nullptrs before deref #6992

zmiller committed:
Actually rename the file itself, not just the config knobs.  Derp.  #7376

zmiller committed:
Merge branch 'V8_9-gt7577-branch'  #7577

GregThain committed:
Check for nullptr deref in secman #6992

GregThain committed:
fstat really can't fail so explicitly ignore error return #6992

markcoatsworth committed:
Added syscall socket + example usage in FileTransfer for credential refresh (#7532)

zmiller committed:
Fix call to slightly oddball interface of krb5_free_addresses()  #6992 #7578

zmiller committed:
Tidy up some debug messages #7577

zmiller committed:
Fix debug message to display name of SchedD #7579

zmiller committed:
Fix debug message to display name of SchedD #7579

zmiller committed:
Merge branch 'V8_8-branch'

zmiller committed:
missed one case of improving debug message #7579

zmiller committed:
Merge branch 'V8_8-branch'

zmiller committed:
Check for NULL value of pointer pw.  #7580

GregThain committed:
Fix dprintf format string error #6992

GregThain committed:
Check return value of get_user_uids #6992

shohamd4 committed:
condor_starter: Support of CPU affinity with docker

If user configured CPU affinity for a given slot, use this CPU affinity
for docker universe jobs running on those slots by using the --cpuset-cpus
option in Docker create cmd API.
Also, made this support mutuale exclusive with the CPU share configuration
(taking into account machineAD number of CPUs), as they have the same effect.

Todd-L-Miller committed:
(#7603)  Allow platforms to include docker:// URLs (URL basename is platform shortname).  Add nmi-build:x86_64_Debian9 to SubmitInfo.pm for stupid technical reasons.

Todd-L-Miller committed:
(#7603)  Allow platforms to include docker:// URLs (URL basename is platform shortname).  Add nmi-build:x86_64_Debian9 to SubmitInfo.pm for stupid technical reasons.

GregThain committed:
Allow stream stdout and stderr to work with files > 32 bits #7547

GregThain committed:
Merge branch 'V8_8-branch'

jasoncpatton committed:
Yank UserUrl out of OAuth keyfile #7463

jasoncpatton committed:
Add common OAuth2 auth/token endpoints to param table #7463

matyasselmeci committed:
Add note about how to add configuration to a minicondor container

timtheisen committed:
Merge branch 'V8_8_8-branch' into V8_8-branch

timtheisen committed:
Merge branch 'V8_9_6-branch'

timtheisen committed:
Merge branch 'V8_8-branch'

johnkn committed:
use initializer lists for classad sink constructor

johnkn committed:
enable consumption policy test on windows, and make it less fragile, fix lookup in consumption policy

JaimeFrey committed:
Set RUNPATH for shared libraries in addition to executables. #7584

Otherwise, RPATH locations are searched before LD_LIBRARY_PATH locations
when libcondor_utils dlopen()s a library (e.g. libssl, gsi).

GregThain committed:
Avoid divide-by-zero in negotiator #7583

GregThain committed:
Merge branch 'V8_8-branch'

GregThain committed:
Remove condor_fix*h files no longer used after std uni removal

Todd-L-Miller committed:
(#7603)  Missed a case where we needed to use $platform_key instead of $platform_name to handle Docker platforms correctly; added abillity to submit tests for Docker platforms.

Todd-L-Miller committed:
(#7603)  Missed a case where we needed to use $platform_key instead of $platform_name to handle Docker platforms correctly; added abillity to submit tests for Docker platforms.

timtheisen committed:
Version history for 8.9.6

timtheisen committed:
Merge branch 'V8_9_6-branch'

timtheisen committed:
Update release date in Manual

timtheisen committed:
Merge branch 'V8_8_8-branch' into V8_8-branch

timtheisen committed:
Update manual date

timtheisen committed:
Merge branch 'V8_9_6-branch'

zmiller committed:
Minor memory leak when using dynamic directories (-d flag to master).  #6992

jasoncpatton committed:
Fix OneDrive auth/token endpoints #7463

zmiller committed:
Fix dynamic directories to not overwrite STARTD_NAME and to not keep expanding on restart.  #7585

zmiller committed:
Merge branch 'V8_8-branch'

markcoatsworth committed:
Added PROVISIONER to parse language (#5622)

markcoatsworth committed:
Merging basic provisioner nodes into master branch (#5622)

johnkn committed:
For #7588, more stable and relable ordering of child-rollup and cross-slot attributes in the Startd.
  These changes are intended to allow for GPUs to assigned to both a primary slot and backfill slot
  with STARTD and PREEMPT policy expressions insuring that only one slot will use the GPU at any one time.
===VersionHistory:Pending===

zmiller committed:
Version History for KERBEROS_SERVER_PRINCIPAL bugfix #7577

johnkn committed:
fix fedora warning for #7588

timtheisen committed:
(#7117) Add passwords.d tokens.d credentials directories to RPM spec

timtheisen committed:
(#7117) Add passwords.d tokens.d credentials directories to .deb

GregThain committed:
Document #7539

GregThain committed:
Merge branch 'master' of /p/condor/repository/CONDOR_SRC

GregThain committed:
Document #7485

timtheisen committed:
(#7569) Fix CA file paths on Debian/Ubuntu

markcoatsworth committed:
Provisioner node assertion bug (#5622)

markcoatsworth committed:
Merge branch 'master' of moria.cs.wisc.edu:/p/condor/repository/CONDOR_SRC

Commit:
fix for #7591: preempting claim in a GPU slot attempts to add another GPU and fails

when preempting, the resources are deleted from the d-slot and then re-added (sigh) for the new job
This hack is incomplete though, when the GPU count is decremented, the assigned GPU is not released,
so a later attempt to bind a GPU into a slot will fail because the count of free GPUs no longer
matches the number of items in the list of free GPUS.  The fix is to unbind the GPU at the same
time that we decrement the GPU count in the d-slot.  Note that this means that the next job can
end up being assigned a different GPU, but since GPUs are currently required to homogeneous, this is ok
for now.

Signed-off-by: John (TJ) Knoeller <johnkn@cs.wisc.edu>

timtheisen committed:
(#7589) Require openssh-server so condor_ssh_to_job works

johnkn committed:
Merge branch 'V8_8-branch' of moria.cs.wisc.edu:/p/condor/repository/CONDOR_SRC into V8_8-branch

GregThain committed:
Document accounting ad attributes #74533

GregThain committed:
Merge branch 'master' of /p/condor/repository/CONDOR_SRC

GregThain committed:
Document #7472

GregThain committed:
Document #7443

johnkn committed:
Merge branch 'V8_8-branch'

johnkn committed:
Merge branch 'master' of moria.cs.wisc.edu:/p/condor/repository/CONDOR_SRC

johnkn committed:
for #7572, allow the argument of the getenv keyword in submit_utils to be a black/white list or a boolean
===VersionHistory:Pending==

Todd-L-Miller committed:
(#7593)  Hoist a syscall out of the close() loop.

JaimeFrey committed:
Remove useless log message about event log not being locked.

Not locking for write is now the default for all job event files, so
this message just clutters the daemon log and raises needless concern
for admins looking through the logs.

JaimeFrey committed:
Fix reading of service account credentials for GCE jobs. #7555

For the new SQL file format used by the GCE client tools, the gce_gahp
wasn't properly reading the access token and expiration from the
secondary file for service credentials.

jasoncpatton committed:
Retry box uploads on broken pipe #7531

timtheisen committed:
Merge branch 'V8_9-gt7505-branch' #7505

timtheisen committed:
(#7469) Add libglobus-gss-assist3 to recommended Debian packages

timtheisen committed:
(#7428) Spruce up manual copyright and fix up man pages

timtheisen committed:
(#7452) Update to scitokens-cpp 0.5.0

timtheisen committed:
Merge branch 'V8_8-branch'

johnkn committed:
turn off pre-release for 8.9.7 release branch

johnkn committed:
change version of master branch to 8.9.8

timtheisen committed:
Revert "Revert "(#7141) Revised python bindings tutorials""

This reverts commit eaa0290e59dc30e1170946a9ec506fbbec35bfe7.

timtheisen committed:
Merge branch 'V8_9_7-branch'

timtheisen committed:
Update version number in manual and add version history section.

timtheisen committed:
Fixup sphinx warning.

GregThain committed:
Use full_write in shadow Remote System Calls #7596

The starter side doesn't check and retry on short writes.
As we're talking over the network, we should loop in the shadow,
not over the network.

GregThain committed:
Merge branch 'V8_8-branch'

GregThain committed:
Document #7596

GregThain committed:
Fix sphinx typo

GregThain committed:
Merge branch 'V8_8-branch'

GregThain committed:
Document #7442

zmiller committed:
Fixed a confusing when history.  #7596

zmiller committed:
Merge branch 'V8_8-branch'

zmiller committed:
Use the directory that conforms to the OAuth and packaging convention.  #7376

timtheisen committed:
Merge branch 'V8_9_7-branch'

JaimeFrey committed:
Use param HAD_USE_PRIMARY for disabling random choice of collector. #7556

JaimeFrey committed:
Docs for HAD_USE_PRIMARY disabling random order of collector query. #7556

JaimeFrey committed:
Docs for HAD_USE_PRIMARY disabling random order of collector query. #7556

zmiller committed:
Change default (and add to param table) MAX_REMAP_RECURSIONS.  #7581

zmiller committed:
Merge branch 'V8_8-branch'

johnkn committed:
have run_test.pl always put the python of the htcondor being tested at the front of the existin PYTHONPATH
this fixes test failures when PYTHONPATH has been set to point to an older system HTCondor

timtheisen committed:
Add sphinx to the container so that we can build the manual

johnkn committed:
add tests to the windows skip list that don't support windows

johnkn committed:
Merge branch 'V8_9_7-branch'

timtheisen committed:
Merge branch 'V8_8-branch' into V8_9_7-branch

markcoatsworth committed:
Fixed DAGMan incorrect warning message (#7548)

markcoatsworth committed:
Added version history (#5622)

markcoatsworth committed:
Added version history (#5622)

markcoatsworth committed:
Added version history (#7337)

markcoatsworth committed:
Merge branch 'V8_9_7-branch', added version history (#7337)

GregThain committed:
Port to cmake 3 #7601

We were accidentally putting condor_ver_info.WISDOM into the list of sources to build libcondorapi.a
and cmake 2.6 was happy with this, but newer cmakes, not so much.

GregThain committed:
Fix warning breaking Fedora build

GregThain committed:
Port condor_tests CMakefile to cmake 3 #7601

cmake 3, in cmake 3 mode, throw an error if you
try to add a dependency on a non-existing target.

The Windows test build was doing this for reasons
described in the comment

timtheisen committed:
Merge branch 'V8_9_7-branch'

GregThain committed:
Skip early out check when surplus code thinks we're over quota conditionally #7602

Todd-L-Miller committed:
(#7603)  Fix cmd_curl_plugin and job_ec2_basic to ignore HTTP_PROXY; it makes no sense in the latter's case anyway, and it breaks things in containers for both.

Todd-L-Miller committed:
(#7603)  Fix cmd_curl_plugin and job_ec2_basic to ignore HTTP_PROXY; it makes no sense in the latter's case anyway, and it breaks things in containers for both.

Todd-L-Miller committed:
(#7603)  For now, just request the same resources for all platforms.

Todd-L-Miller committed:
(#7603)  For now, just request the same resources for all platforms.

markcoatsworth committed:
Added version history (#7414)

brianhlin committed:
Fix a code block formatting issue

brianhlin committed:
Fix a code block formatting issue

jasoncpatton committed:
Move krb and oauth cred dirs out of top level dir #7117

GregThain committed:
More porting to CMake 3 #7601

johnkn committed:
fix crash in token auth, and uninitialized variable in file transfer that lead to URL uploads failing
when there is more than one file.  also fix wrong filename in a log message.

johnkn committed:
fix bug in file_transfer that incorrectly incremented a pointer instead of what it pointed to.

Todd-L-Miller committed:
Merge branch 'master' of moria.cs.wisc.edu:/p/condor/repository/CONDOR_SRC

timtheisen committed:
Merge branch 'V8_9-gt7604a' #7604

timtheisen committed:
Merge branch 'V8_9-gt7604b' #7604

timtheisen committed:
Merge branch 'V8_8-branch' into V8_9_7-branch

timtheisen committed:
Merge branch 'V8_9_7-branch'

Commit:
(#7605) Allocate GPU device name on heap

Signed-off-by: Tim Theisen <tim@cs.wisc.edu>

johnkn committed:
fix uninitialized variable in file transfer ::DoUpload

johnkn committed:
Merge branch 'V8_8-branch' into V8_9_7-branch

timtheisen committed:
Revert "(#7605) Allocate GPU device name on heap"

This reverts commit d2bc94a52187d402cd6ab9a65ce0bbff4d3a42ad.

timtheisen committed:
(#7141) More descriptive names for tutorial sections

johnkn committed:
possible fix for #7605, condor GPU discovery crashes on PowerPC

markcoatsworth committed:
Updated Quick Start Condor Pool with better instructions (#7475)

johnkn committed:
possible fix for #7605, condor GPU utilization crashes on PowerPC

Todd-L-Miller committed:
The extra space here was causing me grief.

Todd-L-Miller committed:
The extra space here was causing me grief.

Todd-L-Miller committed:
Merge branch 'master' of moria.cs.wisc.edu:/p/condor/repository/CONDOR_SRC

johnkn committed:
docs and version history for #7572

zmiller committed:
Versions history for STARTD_NAME issue. #7585

zmiller committed:
Merge branch 'V8_8-branch'

johnkn committed:
version history for #7538 #7486 #7474

GregThain committed:
Revert "Skip early out check when surplus code thinks we're over quota conditionally #7602"

This reverts commit 62ecfc5b648275e0809b60019ab24a35c61baf4d.

GregThain committed:
Lookup Weighted Running and Idle jobs as float #7602

Todd-L-Miller committed:
Fix a typo in the manual.

Todd-L-Miller committed:
Fix a typo in the manual.

Todd-L-Miller committed:
Merge branch 'master' of ssh://ingwe.cs.wisc.edu/p/condor/repository/CONDOR_SRC

Todd-L-Miller committed:
Missed a word on this man page.

Todd-L-Miller committed:
Missed a word on this man page.

Todd-L-Miller committed:
More tiny doc fixes.

Todd-L-Miller committed:
More tiny doc fixes.

Todd-L-Miller committed:
Merge branch 'V8_8-branch' of moria.cs.wisc.edu:/p/condor/repository/CONDOR_SRC into V8_8-branch

timtheisen committed:
Merge branch 'V8_8-branch' into V8_9_7-branch

timtheisen committed:
Merge branch 'V8_9_7-branch'

bbockelm committed:
Filter authentication methods. #7500

- Centralize all method determination logic (which was inconsistent
  based on the call site).
- Remove auth methods from the list that are not in the current build
  (e.g., no NTSSPI on Linux) or cannot be used.
- Remove auth methods that don't exist in any build of HTCondor.
- Ensure we still use the canonical version of 'TOKEN' and 'SCITOKENS'.

Todd-L-Miller committed:
Merge branch 'master' into V8_9-gt7500-branch

Todd-L-Miller committed:
(#7500)  Fix a pair of warnings.

GregThain committed:
Explicitly list source files in condor_utils #7608

GregThain committed:
Clarify the documentation about where  bug reports should go in HTCondor

GregThain committed:
Clarify README.md about prefered places for bug reports.

GregThain committed:
Remove warning #6992

Todd-L-Miller committed:
(#7607)  classad.ExprTree() should fail if its argument has trailing garbage.

Todd-L-Miller committed:
Merge branch 'master' of ssh://ingwe.cs.wisc.edu/p/condor/repository/CONDOR_SRC

Todd-L-Miller committed:
(#7607)  Make the error message less incomprehensible.

GregThain committed:
Merge branch 'master' of /p/condor/repository/CONDOR_SRC

Todd-L-Miller committed:
(#7607)  Release note.

johnkn committed:
version history for #7553, #7521, #7542, #7605, #7360, #7591, #7479

johnkn committed:
fix typos in previous commit

johnkn committed:
Merge branch 'V8_8-branch' into V8_9_7-branch

johnkn committed:
Merge branch 'V8_9_7-branch'

johnkn committed:
version history for #7585 in stable branch

johnkn committed:
Merge branch 'V8_8-branch' into V8_9_7-branch

johnkn committed:
Merge branch 'V8_9_7-branch'

johnkn committed:
version history for #7432

johnkn committed:
version history for #7421 and #7462

johnkn committed:
Add ScratchDirFileCount to job attribute docs. #7486

johnkn committed:
update condor_history man page for #7538

JaimeFrey committed:
Fix broken macos build due to missing include file. #7605

GregThain committed:
Fix assert when using condor_chirp fetch #7612

timtheisen committed:
Merge branch 'V8_8-branch' into V8_9_7-branch

timtheisen committed:
Merge branch 'V8_9_7-branch'

timtheisen committed:
Make curl tests run properly in Docker containers in CHTC

In CHTC, the http_proxy environment variable is set and that interferes
with the curl tests. So, we unset http_proxy for those tests.

GregThain committed:
Remove all CMake calls to selective_glob #7614

This makes rebuilding after someone adds or removes a file
deterministic.

markcoatsworth committed:
Reversed syscall approach to shadow communication (#7532)

markcoatsworth committed:
Merging file transfer plugin enumerated return types into master (#7532)"

Conflicts:
	src/condor_utils/file_transfer.cpp

timtheisen committed:
Add Docker builds for x86_CentOS6 and x86_64_Debian10

timtheisen committed:
Merge branch 'V8_8-branch' into V8_9_7-branch

timtheisen committed:
Merge branch 'V8_9_7-branch'

timtheisen committed:
Fix bad merge. Not building on RHEL6 for development series.

timtheisen committed:
Merge branch 'V8_9_7-branch'

GregThain committed:
Fix incorrect attributes in manual for GROUP_SORT_EXPR

GregThain committed:
Merge branch 'V8_9_7-branch' of /p/condor/repository/CONDOR_SRC into V8_9_7-branch

shamoya commented on src/condor_utils/docker-api.cpp:
```diff
@@ -101,21 +103,34 @@ int DockerAPI::createContainer(
 
-	if (machineAd.LookupInteger(ATTR_CPUS, cpus)) {
-		cpuShare = 100 * cpus;
+	if (affinity_mask)
```
Coding style is to keep "{" after expression.

shamoya commented on src/condor_utils/docker-api.cpp:
```diff
@@ -101,21 +103,34 @@ int DockerAPI::createContainer(
 	// Configure resource limits.
-	
-	// First cpus
-	int  cpus;
```
My only question is: do the "cpuset-cpus" and "cpu-shares" features are mutually exclusive?
Cause we treat them that way. if not we can leave the the previous code (cpu-shares) as-is and just add the cpu-shares code block.

shamoya commented on src/condor_starter.V6.1/docker_proc.cpp:
```diff
@@ -182,9 +182,11 @@ int DockerProc::StartJob() {
 
 	buildExtraVolumes(extras, *machineAd, *JobAd);
 
+	int *affinity_mask = makeCpuAffinityMask(Starter->getMySlotNumber());
```
So as I understand, makeCpuAffinityMask will check if ENFORCE_CPU_AFFINITY is set and if so it will set the affinity_mask (it malloc it, so I guess we also need to free it after calling CreateContainer).
So it's the same config_val for both regular execution and docker. I think it's ok, or you think we need a new dedicated config for CPU_AFFINITY_DOCKER ... ?

shamoya reviewed (commented):
Also I went over the commits they use, we need to add some info in the commit message.
Maybe just mention the configuration macros  used to set the affinity and if it's mutuale excluse with the cpu-shares

timtheisen committed:
Tune BaTLab Docker job resource request for CHTC

timtheisen committed:
Merge branch 'V8_8-branch' into V8_9_7-branch

timtheisen committed:
Merge branch 'V8_9_7-branch'

GregThain committed:
Only built condor_version.cpp once #7614

brianhlin committed:
Fix missing bosco/glite/libexec dir on the remote host

brianhlin committed:
Fix missing bosco/glite/libexec dir on the remote host

johnkn committed:
fix #7609 python queue_with_itemdata segfault

johnkn committed:
Merge branch 'V8_8-branch' into V8_9_7-branch

johnkn committed:
Merge branch 'V8_9_7-branch'

GregThain committed:
Remove unused cmake macros #7614

GregThain committed:
Add missing file from last commit #7614

timtheisen committed:
Increase request_memory and add x86_64_CentOS8 platform

timtheisen committed:
Merge branch 'V8_8-branch' into V8_9_7-branch

timtheisen committed:
Merge branch 'V8_9_7-branch'

GregThain committed:
Remove condor_glob from cmakefiles, list files explicitly #7614

markcoatsworth committed:
Fixed mangled merge (#7532)

markcoatsworth committed:
Merge branch 'master' of moria.cs.wisc.edu:/p/condor/repository/CONDOR_SRC

GregThain committed:
Document AssignAccountingGroup metaknob

GregThain committed:
Fix sphinx syntax error in condor_submit man page

GregThain committed:
Clean up docs around AssignAccountingGroup metaknob

JaimeFrey committed:
Fix policy in security sessions sent by schedd to negotiator/collector #7615

When the schedd creates a non-negotiated security session for use by the
collector and negotiator to contact it, the policy it sends out doesn't
include the encryption method or expiration time. This caused breakage
when the default encryption method changed in 8.9.4 (8.9.3 and 8.9.4+
could not talk with each other).

tannenba committed:
Fix problems with ssh_to_job to container if BIN is not /usr/bin. #7613

Note we now pass the value of config knob BIN to the job via the
environment variable _CONDOR_BIN.

tannenba committed:
Documentation and version history for #7613.

tannenba committed:
Merge branch 'V8_8-branch' of ssh://ingwe.cs.wisc.edu/p/condor/repository/CONDOR_SRC into V8_8-branch

johnkn committed:
fix for #7438, POOL_HISTORY_MAX_STORAGE doesn't work on Windows

johnkn committed:
for #7616 condor_q -submitters now correctly queries for jobs that match the submitter name

GregThain committed:
Replace condor_daemon macro with globless function with named parameters #7614

timtheisen committed:
Increase request memory from 3G to 4G for Docker containers

johnkn committed:
add SUBMIT_GENERATE_CUSTOM_RESOURCE_REQUIREMENTS to disable submit requirements generation for job attriburtes that start with Request

johnkn committed:
Merge branch 'V8_8-branch' into V8_9_7-branch

johnkn committed:
Merge branch 'V8_9_7-branch'

johnkn committed:
turn off pre-release for 8.8.9 release branch

johnkn committed:
Merge branch 'V8_8_9-branch' into V8_9_7-branch

johnkn committed:
Merge branch 'V8_9_7-branch'

johnkn committed:
set version of V8_8-branch to 8.8.10

johnkn committed:
Merge branch 'V8_8-branch'

timtheisen committed:
(#7520) Minor refactor

timtheisen committed:
Get RPM builds to work on CentOS8 by removing -g3

timtheisen committed:
Merge branch 'V8_9-gt7618-branch' #7618

timtheisen committed:
(#7362) This was a one off tarball. Not compatible with general release.

timtheisen committed:
Merge branch 'V8_8_9-branch' into V8_9_7-branch

timtheisen committed:
Update manual for 8.8.10

timtheisen committed:
Fixup sphinx-build warning

timtheisen committed:
Merge branch 'V8_8_9-branch' into V8_8-branch

timtheisen committed:
Merge branch 'V8_9_7-branch'

timtheisen committed:
Merge branch 'V8_8-branch'
